lib/spack/llnl/util/cpu/microarchitecture.py

```diff
@@ -216,32 +216,18 @@ def optimization_flags(self, compiler, version):
             compiler_info = [compiler_info]
 
         def satisfies_constraint(entry, version):
-            min_version, max_version = entry['versions'].split(':')
-
-            # Check version suffixes
-            min_version, _, min_suffix = min_version.partition('-')
-            max_version, _, max_suffix = max_version.partition('-')
-            version, _, suffix = version.partition('-')
-
-            # If the suffixes are not all equal there's no match
-            if suffix != min_suffix or suffix != max_suffix:
-                return False
-
-            # Assume compiler versions fit into semver
-            tuplify = lambda x: tuple(int(y) for y in x.split('.'))
-
-            version = tuplify(version)
-            if min_version:
-                min_version = tuplify(min_version)
-                if min_version > version:
-                    return False
-
-            if max_version:
-                max_version = tuplify(max_version)
-                if max_version < version:
-                    return False
-
-            return True
+            constraints = llnl.util.version.VersionList()
```

Comment:
We've taught users that they can create arbitrary versions strings for Spack compilers as a way of tracking which compilers they have configured with which options. Those compilers break this code as-is, and I don't think that's a feature we can easily move away from.

The goal with this PR was to refactor version into something external that `cpu` could continue to depend on after it has been factored out of Spack as an external.

Reply:
**TL;DR** I completely agree fixing the handling of custom string versions. What I'm saying is that we should do it in a way that doesn't make `archspec.cpu` depend on `spack.version`. Looking better at the implications, we might even add a new type of bugs for custom versions doing it this way  - more details below. What I think we should do instead is:

1. Improve `archspec.cpu` API to report if a version is conforming to the expected structure
2. Handle custom version strings for compilers in Spack, possibly in `build_environment.py`

Below a broader explanation, let me know what you think.

----

> We've taught users that they can create arbitrary versions strings for Spack compilers as a way of tracking which compilers they have configured with which options. Those compilers break this code as-is, and I don't think that's a feature we can easily move away from.

I completely agree fixing this and I'm not suggesting we move away from supporting string versions for compilers in Spack.

> The goal with this PR was to refactor version into something external that cpu could continue to depend on after it has been factored out of Spack as an external

This is more problematic, and in my opinion not the best way to deal with the issue. If we want to treat `archspec.cpu` as a separate package there's absolutely no benefit adding a dependency on `spack.version`. There are at least two big downsides though:

1. The peculiar way in which we treat versions in Spack won't work well with the information we have in `microarchitectures.json`. Since we make use of the version information to map to the correct set of optimization flags this might cause troubles and make the program fail or behave incorrectly. As a practical example `gcc@system` will be always compared less than any numeric version and `gcc@develop` always compared greater than any numeric version. **This means we'll disregard the underlying version of `gcc` and take the wrong decision in both cases.** In the end this might bring unexpected behaviors (demotion of architectures) or bugs (I made a `gcc@9.2.0` always fail at configure time with this PR by calling it `gcc@foo` as it was using deprecated flags).

2. Even if using `spack.version` would result in correct behavior, adding a dependency on it will surely make `archspec.cpu` harder to adopt. For a user it means that they agree on treating versions the same way Spack does, and they agree depending on another package for doing so. For us it means packaging `spack.version` as an external package before starting the work on `archspec`. Since version checking is only required in a single method of a class I would stay away on adding a dependency for that.

What I believe we should do instead is:

1. Improve `archspec.cpu` API to report if a version is conforming to the expected structure
2. Handle custom version strings for compilers in Spack, possibly in `build_environment.py`

My idea is that if we can change this code:

https://github.com/spack/spack/blob/ea4f9ebbde50ae3606a6d8955512bc1b54299c51/lib/spack/spack/build_environment.py#L203-L205

into something like:

```python
compiler_version = compiler.version
if not cpu.is_valid_compiler_version(compiler_version):
    # Try to detect the underlying compiler version in some way, maybe using spack.compiler
    compiler_version = ...

isa_arg = spec.target.optimization_flags(compiler.name, compiler_version)
```
we would solve our issue in a neater way. What do you think?

Reply:
Yes, I think that's a reasonable idea.

We could add a field to the compiler config `version_number` that users are not free to change. That way we wouldn't always have to reconstruct the version number at run time. And if `version_number` isn't set we could use the same detection we use at `compiler find` time.

If we make `version_number` a caching variable on `spack.compiler.Compiler`, we could also use it for the version checks in properties like `cxx11_flag` to avoid the buggy behavior we get there with arbitrary version strings.